apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -1520,78 +1568,70 @@ void RetryAcksAndExtends(bool? ack, AckError ackError, IEnumerable<TimedId> exte
                     }
 
                     var idsToRetry = hasAcksOrNacks ? ackOrNackIds : extendIds.Select(j => j.Id);
-                    var retryableErrorIds = new ConcurrentBag<string>();
+                    var retryableErrorIds = new List<string>();
                     // Not all temporaryFailures may need to be retried.
                     // See if the id exists in the _retryableIds dictionary and only then retry.
-                    foreach (var id in idsToRetry.Where(j => _retryableIds.ContainsKey(j)))
-                    {
-                        retryableErrorIds.Add(id);
-                    }
+                    retryableErrorIds.AddRange(idsToRetry.Where(j => _retryableIds.ContainsKey(j)));
 
                     if (retryableErrorIds.Count == 0)
                     {
                         // Nothing to retry.
                         return;
                     }
 
-                    // All ids in retryableErrorIds will have the same backoff.
-                    // Get the retryInfo for the first retryableErrorId from the _retryableIds dictionary.
-                    if (!_retryableIds.TryGetValue(retryableErrorIds.First(), out var retryInfo))
+                    foreach (var id in retryableErrorIds)
                     {
-                        // Nothing to retry.
-                        return;
-                    }
+                        // Only ids that exist in _retryableIds dictionary are added to retryableErrorIds from idsToRetry.
```

Comment:
And we won't ever update `_retriableIds` in a different thread, between the above code and this? I *think* that's the case, but please confirm if that's your understanding as well. (I believe the individual client is effectively single-threaded, just highly asynchronous. So the fact that there aren't any awaits involved means we should be fine. Is that right?)

Reply:
Right Jon. This is my understanding as well. I was trying to be defensive just to be safe.